spacelift/resource_role_attachment.go

```diff
@@ -0,0 +1,248 @@
+package spacelift
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal"
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal/structs"
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal/validations"
+)
+
+const (
+	apiKeyRoleAttachmentPrefix          = "API"
+	idpGroupMappingRoleAttachmentPrefix = "IDP"
+)
+
+func resourceRoleAttachment() *schema.Resource {
+	return &schema.Resource{
+		Description: "" +
+			"`spacelift_role_attachment` represents a Spacelift role attachment " +
+			"between an API key and a role; or between an IdP Group Mapping and a role.\n" +
+			"Either `api_key_id` or `idp_group_mapping_id` must be set, but not both.",
+
+		CreateContext: resourceRoleAttachmentCreate,
+		ReadContext:   resourceRoleAttachmentRead,
+		DeleteContext: resourceRoleAttachmentDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"api_key_id": {
+				Type:         schema.TypeString,
+				Description:  "ID of the API key to attach to the role (ULID format). For example: `01F8Z5K4Y3D1G2H3J4K5L6M7N8`.",
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"api_key_id", "idp_group_mapping_id"},
+			},
+			"idp_group_mapping_id": {
+				Type:        schema.TypeString,
+				Description: "ID of the IdP Group Mapping to attach to the role (ULID format). For example: `01F8Z5K4Y3D1G2H3J4K5L6M7N8`.",
+				Optional:    true,
+				ForceNew:    true,
+			},
+			"role_id": {
+				Type:             schema.TypeString,
+				Description:      "ID of the role to attach to the API key (ULID format). For example: `01F8Z5K4Y3D1G2H3J4K5L6M7N8`.",
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validations.DisallowEmptyString,
+			},
+			"space_id": {
+				Type:             schema.TypeString,
+				Description:      "ID of the space where the role attachment should be created",
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validations.DisallowEmptyString,
+			},
+		},
+	}
+}
+
+func resourceRoleAttachmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiKeyID := d.Get("api_key_id").(string)
+
+	if apiKeyID != "" {
```

Comment:
If your intention is to prep the code structure for the upcoming few actors, like users, user groups, etc . , maybe a switch would be a better approach here?